### ISSUE TITLE
"Apply In Editor" Very slow for simple requests and applies code incorrectly

### ISSUE THREAD

**Issue by AlfredoDinunzioTR** [microsoft/vscode-copilot-release#3085]
Sorry if this is a duplicate, I did search and could not find it.   I don't think it is a new issue but as "Copilot Edits" continues to get better, I am hoping that the regular chat applying can also be improved. 

- Copilot Chat Extension Version:  v0.23.2024102903 (pre-release)
- VS Code Version: Version: 1.95.3 
- OS Version: Windows 11
- Logs: I could not find logs that mentioned anything about this. 

Steps to Reproduce:

1. Suggest change to simple code block (in my case, 2 lines)
2. Chat returns a change that modifies those 2 lines and adds 2 more lines
3. Click "Apply in Editor"
4. Counted ~45 Seconds of loading before anything happens
5. Once it applied, it only modified the 2 lines and did not enter the 2 new lines. 

I would expect all 4 lines to be applied, not only the 2 existing ones. I also don't think it should take that long to apply when the only context given was ~8 lines.

**Comment by alexdima** [comment]
@AlfredoDinunzioTR Do you remember, was the initial file very large?

**Comment by VanderSP** [comment]
this happens to me, maybe code base larger? because the file actually small

**Comment by Shaerif** [comment]
You're right, this shouldn't happen.

**Comment by VanderSP** [comment]
Microsoft sleeps fast

**Comment by JoseKook** [comment]
Hola, tengo el mismo problema, lograron solucionarlo?

**Comment by DGLcsGaming** [comment]
Same issue for me, if the code is already generated in the chat window, why does it take ages for it to be applied ?

**Comment by yllekz** [comment]
I feel this has gotten significantly and exponentially worse in recent times. I often have to click "apply in editor" and wait upwards of 5 to 10 minutes before it completes. It also appears that the action is attempting to rewrite the entire file each time as opposed to just the localized changes that it pre-identified in the chat. This cannot be an acceptable metric of performance.

**Comment by Zoleaf** [comment]
same problem at our side

**Comment by DGLcsGaming** [comment]
I read somewhere previously from one of the maintainers of `vscode-copilot-release issues` that they're using the LLM to decide where to modify the code, but this seems very inefficient. If this project was opensource I'm sure will would create an algorithm to just that on the client side, so let's just hope the vscode copilot team address this issue 🤲

**Comment by hmorneau** [comment]
@aeschli @joyceerhl #3629 Can we get an update on this. It's been open for a while, there was some work being done on this last month, since then, nothing. It's probably one, if not the biggest pain point people report right now while using GH Copilot.

**Comment by martin-wintz** [comment]
This is incredibly painful. I was forced to switch from Cursor to copilot at work and this issue renders the tool nearly useless. At this point I can write the code faster myself.

**Comment by YinAoXiong** [comment]
same problem

**Comment by maurert** [comment]
I'm using VS-Code with Copilot (Claude 3.7 Sonnet Thinking). The AI gives an answer in the chat within a resonable length of time, but when I hit the apply button, the merge takes a very long time.

**Comment by smissu** [comment]
> I'm using VS-Code with Copilot (Claude 3.7 Sonnet Thinking). The AI gives an answer in the chat within a resonable length of time, but when I hit the apply button, the merge takes a very long time.

I switch to the GPT-4o to apply the code that was generated from Claude 3.7 and the updates apply a bit faster. Not ideal to switch ever  time but better than waiting on 3.7 to make the updates.

**Comment by hmorneau** [comment]
Every time I see an upgrade of vs code insider I eagerly try the apply feature to see if the new improved fast version have been rolled out.

**Comment by smissu** [comment]
> Every time I see an upgrade of vs code insider I eagerly try the apply feature to see if the new improved fast version have been rolled out.

Would be nice if copilot allowed you to specify a model just for writing the code as well as a model for generating code.

**Comment by hmorneau** [comment]
> Would be nice if copilot allowed you to specify a model just for writing the code as well as a model for generating code.

Wrong solution to this problem. Cursor solved this with a custom model that do 5000 tokens/sec. Why would anyone select any other models? They just need one good model for the task and use that model.

**Comment by AlfredoDinunzioTR** [comment]
> [@AlfredoDinunzioTR](https://github.com/AlfredoDinunzioTR) Do you remember, was the initial file very large?

Hello, sorry for the very delayed response. Given the recent activity on this issue, I wanted to provide an answer and share some updated observations. While I don't recall the exact file that prompted me to create this issue, it was likely around 1000 lines of code. 

Based on my continued experience with Copilot:

- The "Apply in editor" functionality appears to scan the entire file regardless of change size, with processing time scaling with file size rather than change complexity
- On a recent example with a 1500-line file, a simple comment addition around line 900 took approximately 4 minutes to complete, with roughly the first 2 minutes spent just scrolling to the change location
- The scanning process has improved in speed but remains significantly slower than expected
- I've noticed fewer mistakes in the applied code compared to when I first reported this issue
- I now use "Insert at Cursor" more frequently for exact replacements, reserving "Apply to editor" for cases where Copilot has omitted portions of a function due to length constraints

**Comment by AreAArseth** [comment]
I have also seen this issue around 1000 lines. Also in Agent mode it sometimes makes multiple passes on the same file changing 2-5 lines. Makes everything super slow.

**Comment by kexul** [comment]
Any update?

**Comment by pbartoszek** [comment]
The same problem with slow apply on my side as well. My file is around 150 lines only and it takes like 60s seconds.
